Create 2025-03-12-solidity-0.8.29-release-announcement.md (#129)
ethereum/solidity-website: src/posts/2025-03-12-solidity-0.8.29-release-announcement.md

This latest version of the compiler brings support for EVM Object Format (experimental), Storage Layout Specifiers, and more! 

## Notable Features

### Experimental support for EVM Object Format

0.8.29 brings us experimental support for EVM Object Format (EOF).
Please note that the feature can only be enabled when compiling for the Osaka EVM version, which has not yet been deployed to mainnet or testnets.

The experimental backend is still incomplete and implements only a subset of EOF features.
This subset, however, is already sufficient to support compiling arbitrary Solidity contracts.
The missing elements are, for the most part, opcodes like `RJUMPV` or `EXCHANGE`, which will let us generate more efficient code in the future, but were not strictly necessary to be able to generate working bytecode.

Note that existing contracts may not compile without adjustments, due to some features no longer existing on EOF.
For example, gas and code introspection are no longer possible, which means that it is impossible to generate code for the `gas` call option on external calls or builtins like `gasleft()` and `<address>.code`.

> The changes to contract creation also affect the operator`new`, which now always uses a `salt`.

Reviewer: ```suggestion
The changes to contract creation also affect the operator `new`, which now always uses a `salt`.
```

We would like to thank the [Ipsilon](https://notes.ethereum.org/@ipsilon/about) team for their amazing work on EOF, and in particular [Radek Zagórowicz](https://github.com/rodiazet) for his work on the Solidity implementation.

### Support for Storage Layout Specifiers

This release introduces support for allowing the relocation of contract storage to an arbitrary location.

The support for specifying storage locations for storage variables has been one of the most discussed issues.
It only gained more popularity and became a critical issue in the community with [EIP-7702](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7702.md) and the growing popularity of account abstraction.

Based on feedback and discussions, we shared a proposal for explicit storage layout syntax with various possible approaches and selected a candidate for implementation among the variants which has now been implemented in this release.

Note that storage layout specifiers currently need to be literals. This limitation will be relaxed to allow more expressions in future releases.

> ### Initial ethdebug

Reviewer 2: ```suggestion
### Initial ethdebug support
```
maybe?

> This release also introduces the first experimental implementation to support [ethdebug](https://ethdebug.github.io/format/spec/overview) - a debugging data format suitable for smart contracts. 

Reviewer 2: ```suggestion
This release also introduces the first experimental step towards supporting [ethdebug](https://ethdebug.github.io/format/spec/overview) - a debugging data format suitable for smart contracts. 
```

Might suggest that we have all that is in the spec already, doesn't it? Also fine either way, though.

0.8.29 brings us experimental support for EVM Object Format (EOF).
Please note that the feature can only be enabled when compiling for the Osaka EVM version, which has not yet been deployed to mainnet or testnets.

The experimental backend is still incomplete and implements only a subset of EOF features.
This subset, however, is already sufficient to support compiling arbitrary Solidity contracts.
The missing elements are, for the most part, opcodes like `RJUMPV` or `EXCHANGE`, which will let us generate more efficient code in the future, but were not strictly necessary to be able to generate working bytecode.

Note that existing contracts may not compile without adjustments, due to some features no longer existing on EOF.
For example, gas and code introspection are no longer possible, which means that it is impossible to generate code for the `gas` call option on external calls or builtins like `gasleft()` and `<address>.code`.
The changes to contract creation also affect the operator `new`, which now always uses a `salt`.
The `salt` has a default value (`0`), but will soon become mandatory, since using the same value with the same contract results in the same deployment address 

> (unlike `new` outside of EOF, which uses account's nonce to generate a unique address for every deployment).

Reviewer: ```suggestion
(unlike `new` outside of EOF, which uses an account's nonce to generate a unique address for every deployment).
```

> Finally, many opcodes were removed or replaced, in particular the external calls, so contracts making heavy use of inline assembly will likely not compile as is.

Reviewer: ```suggestion
Finally, many opcodes were removed or replaced, in particular the external calls, so contracts making heavy use of inline assembly will likely not compile as-is.
```

This should alleviate any possible code size regressions users may still see in the current state of the implementation.

Please keep in mind that at the time of this release [the scope of the Fusaka upgrade](https://eips.ethereum.org/EIPS/eip-7607) has not yet been finalized and the EOF specification was still undergoing changes in the recent months.
Our implementation does not account for these changes yet and was based on [the unified EOF specification as of 2024-09-05](https://github.com/ipsilon/eof/blob/7728efe7433b8521c8b6711610e56c6122440795/spec/eof.md) and tested against [evmone 0.13.0](https://github.com/ethereum/evmone/releases/tag/v0.13.0).

To try it out with your contract, use `--experimental-eof-version 1` on the command line or `settings.eofVersion: 1` in Standard JSON and remember to select an EVM version that supports it (`--evm-version osaka`/`settings.evmVersion: "osaka"`).

We would like to thank the [Ipsilon](https://notes.ethereum.org/@ipsilon/about) team for their amazing work on EOF, and in particular [Radek Zagórowicz](https://github.com/rodiazet) for his work on the Solidity implementation.

> ### Support for Storage Layout Specifiers

Author: Should we name the title as `Support for Custom Storage Layouts` for consistency?

> ### Support for Storage Layout Specifiers

Reviewer 3: Sure. I thought the titles could be more concise but if you're doing that intentionally to keep consistent style that's fine with me too.

Currently the syntax is very limited: the base location can only be a literal expression and applies to the whole inheritance tree.
As the next step we are planning to allow using constants and add special-purpose helpers for the most commonly used layouts (in particular [ERC-7201: Namespaced Storage Layout](https://eips.ethereum.org/EIPS/eip-7201)).
Eventually we may support the full range of features described in the [initial syntax proposal](https://notes.ethereum.org/@solidity/explicit-storage-layout-syntax), though we want to do that gradually to avoid overcomplicating it and address the most relevant use cases first.
The community feedback will drive the future of this feature, so do not hesitate to share your opinion on [our forum](https://forum.soliditylang.org) and tell us which aspects are the most important for your contract.

To learn more about the new feature, please consult the [custom storage layout documentation](https://docs.soliditylang.org/en/v0.8.29/contracts.html#custom-storage-layout).

> ### Initial ethdebug Support

Reviewer 3: We might also want to mention ethdebug in the initial blurb because that's the most visible part when this is posted on socials.

> ### Initial ethdebug Support

Author: I had made that change while I was editing only to realise that I hadn't committed it while you were reviewing. So some of these suggestions are either outdated or have been addressed in my commit. Will make sure it shows up in the final version :)

> ### Initial ethdebug Support

Author: BTW, in my final commit, I have changed the title to `Initial Support for ethdebug` to stay consistent with our format both in terms of capitalisation in titles as well as naming them - most of them across all posts are in the format `['experimental' or 'initial' if applicable] Support for <name of the feature>`

> ### Initial ethdebug Support

Reviewer 3: Sure. Sounds good.

This release also introduces the first experimental step towards supporting [ethdebug](https://ethdebug.github.io/format/spec/overview) - a debugging data format suitable for smart contracts. The ethdebug specification is still under development. Join the [Matrix.chat](https://matrix.to/#/#ethdebug:matrix.org) or watch the [GitHub repo](https://github.com/ethdebug/format) to follow the ongoing development. There is a ethdebug meeting every second week on Thursdays at 17:00 Berlin time. Meetings are announced in the [Matrix.chat](https://matrix.to/#/#ethdebug:matrix.org).

> Right now the current implementation supports the generation of [instructions and source ranges](https://ethdebug.github.io/format/spec/program). This initial implementation is not supporting the use of the optimizer, is ony available with the via-ir pipeline and is still missing many important features.

Author: I believe the use of "the" here with generation is the correct usage. :)

> If you want to give it a try, you can generate ethdebug output by using the CLI: `solc --via-ir --ethdebug|--ethdebug-runtime [..]`. To request the ethdebug artifacts in standard json, select `evm.bytecode.ethdebug` and/or `evm.deployedBytecode.ethdebug` - remember that `viaIR` need to be set.

Reviewer: ```suggestion
If you want to give it a try, you can generate ethdebug output by using the CLI: `solc --via-ir --ethdebug|--ethdebug-runtime [..]`. To request the ethdebug artifacts in standard json, select `evm.bytecode.ethdebug` and/or `evm.deployedBytecode.ethdebug` - remember that `viaIR` needs to be set.
```

To learn more about the new feature, please consult the [custom storage layout documentation](https://docs.soliditylang.org/en/v0.8.29/contracts.html#custom-storage-layout).

### Initial Support for ethdebug

This release also introduces the first experimental step towards supporting [ethdebug](https://ethdebug.github.io/format/spec/overview) - a debugging data format suitable for smart contracts.

The ethdebug specification is still under development.
Join the [ethdebug Matrix channel](https://matrix.to/#/#ethdebug:matrix.org) or watch the [ethdebug repository on Github](https://github.com/ethdebug/format) to follow the ongoing development.
There is an ethdebug meeting every second week on Thursdays at 17:00 Berlin time.
Meetings are announced on the Matrix channel.

The current implementation supports the generation of [instructions and source ranges](https://ethdebug.github.io/format/spec/program).
This initial version only supports unoptimized compilation via IR and is still missing many important features.

> If you want to give it a try, you can enable the ethdebug output on the command line with `--ethdebug`/`--ethdebug-runtime`.
> 
> To request the ethdebug artifacts in Standard JSON, add `"evm.bytecode.ethdebug"`/`"evm.deployedBytecode.ethdebug"` in `settings.outputSelection` (note that the `"*"` wildard does *not* enable it).
> 
> Please also remember that `settings.viaIR: true`/`--via-ir` is necessary for the feature to work.

Reviewer 3: ```suggestion
If you want to give it a try, you can enable the ethdebug output on the command line with `--ethdebug`/`--ethdebug-runtime`.
To request the ethdebug artifacts in Standard JSON, add `"evm.bytecode.ethdebug"`/`"evm.deployedBytecode.ethdebug"` in `settings.outputSelection` (note that the `"*"` wildard does *not* enable it).
Please also remember that `settings.viaIR: true`/`--via-ir` is necessary for the feature to work.
```
